zypp/misc/TestcaseSetup.h

```diff
@@ -0,0 +1,132 @@
+/*---------------------------------------------------------------------\
+|                          ____ _   __ __ ___                          |
+|                         |__  / \ / / . \ . \                         |
+|                           / / \ V /|  _/  _/                         |
+|                          / /__ | | | | | |                           |
+|                         /_____||_| |_| |_|                           |
+|                                                                      |
+\---------------------------------------------------------------------*/
+/** \file	zypp/misc/TestcaseSetup.h
+ *
+*/
+
+#ifndef ZYPP_MISC_TESTCASESETUP_H
+#define ZYPP_MISC_TESTCASESETUP_H
+
+#include <zypp/Arch.h>
+#include <zypp/Locale.h>
+#include <zypp/Pathname.h>
+#include <zypp/ResolverFocus.h>
+#include <zypp/Url.h>
+#include <zypp/base/PtrTypes.h>
+#include <zypp/base/SetTracker.h>
+#include <zypp/sat/Queue.h>
+#include <zypp/target/modalias/Modalias.h>
+
+#include <optional>
+#include <vector>
+
+namespace zypp {
+  class RepoManager;
+}
+
+namespace zypp::misc::testcase
+{
+
+  enum class TestcaseRepoType {
+    Helix,
+    Testtags,
+    Url
+  };
+
+  struct RepoDataImpl;
+  struct ForceInstallImpl;
+  struct TestcaseSetupImpl;
+
+  class RepoData {
+  public:
+    RepoData ();
+    RepoData ( RepoDataImpl &&data );
+    TestcaseRepoType type() const;
+    const std::string &alias() const;
+    uint priority() const;
+    const std::string &path() const;
+
+    const RepoDataImpl &data() const;
+    RepoDataImpl &data();
+  private:
+    RWCOW_pointer<RepoDataImpl> _pimpl;
+  };
+
+  class ForceInstall {
+  public:
+    ForceInstall ();
+    ForceInstall ( ForceInstallImpl &&data );
+    const std::string &channel () const;
+    const std::string &package () const;
+    const std::string &kind () const;
+
+    const ForceInstallImpl &data() const;
+    ForceInstallImpl &data();
+  private:
+    RWCOW_pointer<ForceInstallImpl> _pimpl;
+  };
+
+  class TestcaseSetup
+  {
+  public:
+
+    TestcaseSetup();
```

Comment:
When using a pimpl I've got the habit to declare a DTOR (defined together with the CTOR, even if empty).
This prevents g++ from sometimes complaining about having to destroy the incomplete Impl class when a Class instance gets deleted (if the default dtor is used). Same applies to RepoData and ForceInstall.

Reply:
Oh, yes very good point. I just forgot it because it did not complain this time, for whatever reason... --> Fixed